app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js

```diff
@@ -194,27 +201,50 @@ export async function switchChain(
   networkClientId,
   approvalFlowId,
   {
-    getChainPermissionsFeatureFlag,
     setActiveNetwork,
     endApprovalFlow,
     requestUserApproval,
     getCaveat,
-    requestPermittedChainsPermission,
+    requestPermissionApprovalForOrigin,
+    updateCaveat,
   },
 ) {
   try {
-    if (getChainPermissionsFeatureFlag()) {
-      const { value: permissionedChainIds } =
-        getCaveat({
-          target: PermissionNames.permittedChains,
-          caveatType: CaveatTypes.restrictNetworkSwitching,
-        }) ?? {};
-
-      if (
-        permissionedChainIds === undefined ||
-        !permissionedChainIds.includes(chainId)
-      ) {
-        await requestPermittedChainsPermission([chainId]);
+    const caip25Caveat = getCaveat({
+      target: Caip25EndowmentPermissionName,
+      caveatType: Caip25CaveatType,
+    });
+
+    if (caip25Caveat) {
+      const ethChainIds = getPermittedEthChainIds(caip25Caveat.value);
+
+      if (!ethChainIds.includes(chainId)) {
+        if (caip25Caveat.value.isMultichainOrigin) {
+          return end(
+            new Error(
+              'cannot switch to chain that was not permissioned in the multichain flow',
+            ),
+          ); // TODO: better error
+        }
+        await requestPermissionApprovalForOrigin({
+          [PermissionNames.permittedChains]: {
+            caveats: [
+              { type: CaveatTypes.restrictNetworkSwitching, value: [chainId] },
+            ],
+          },
+        });
+
+        const updatedCaveatValue = addPermittedEthChainId(
+          caip25Caveat.value,
+          chainId,
+        );
```

Comment:
we need to setAccounts on this again, or the added scope will have no accounts. This only matters if permittedChains can be removed though, and that doesn't seem to be the case

Reply:
leaving this as is

Reply:
> we need to setAccounts on this again, or the added scope will have no accounts. This only matters if permittedChains can be removed though, and that doesn't seem to be the case

I don't think I'm following this. Could you elaborate?

Reply:
Shouldn't we have to add accounts? Not understanding why 

> This only matters if permittedChains can be removed though

Reply:
> Shouldn't we have to add accounts? Not understanding why

At this point in code, a CAIP-25 permission exists. A CAIP-25 permission can only exist if accounts and a permitted chain were granted. This means that there's at least one scope with all the eth accounts permitted assigned to it in the CAIP-25 permissions caveat value. If this was multichain execution, then yes it would matter that the accounts are not assigned to the newly permitted chain, but we're disallowing modification of CAIP-25 permissions granted from the multichain flow here, so this permission can only work in the EIP-1193 flow.

It does not matter in legacy/EIP-1193 because we get "eth_accounts" from a CAIP-25 permission by getting the unique set of all eip155 accounts on all eip155 scope objects. As long as the full set of accounts exists on one of the scopeObjects then it exists. Does that make sense?

Reply:
Got it! Makes sense! Thanks for the explanation!

Reply:
🤔 I see your point. I think we should revisit this subsequently. I don't love that the permissions state involves implicit (and invisible) permissions on top of them but I suppose we do have a flag (`isMultichainOrigin`) that atleast clues us when different rules are operative.